packages/cli/src/register-command.ts

```diff
@@ -0,0 +1,220 @@
+import inquirer from 'inquirer';
+import { Command } from 'commander';
+
+import { Choice } from './prompts/choices';
+import Telemetry from './telemetry';
+
+export function generateSignature(name: string, args: Arg[]): string {
+  return [name, ...args.map(a => `[${a.name}${a.variadic ? '...' : ''}]`)].join(' ');
+}
+
+export interface Question {
+  message: string;
+  choices?: Choice[];
+  validate: (value: string) => boolean;
+  validationError?: string;
+}
+
+// Param is the things common to both positional arguments (Arg) and options (Option).
+// TODO: Validate that only the last positional argument is variadic.
+type Param = ParamSimple | ParamVariadic;
+
+interface ParamSimple {
+  variadic?: false;
+  prompt?: (argsAndOpts: ArgsAndOpts) => Promise<Question>;
+}
+
+interface ParamVariadic {
+  variadic: true;
+  prompt?: (argsAndOpts: ArgsAndOpts) => Promise<Question[]>;
+}
+
+export type Arg = Param & { name: string };
+
+export interface Option extends ParamSimple {
+  format: string;
+  description: string;
+  default?: string | boolean;
+}
+
+// This was added to support aborting deploy and running create instead. We
+// need to abort in the preAction so that the interactive prompts for deploy
+// are not presented to the user, because create manages its own interactivity.
+export class AbortAction {
+  constructor(readonly callback: () => Promise<void>) {}
+}
+
+export function register(program: Command, spec: CommandSpec, getAction: () => Promise<Action>): void {
+  const signature = generateSignature(spec.name, spec.args);
+
+  const command = program
+    .command(signature)
+    .description(spec.description)
+    .action(async (...actionArgs: unknown[]) => {
+      const { preAction, action }: Action = await getAction();
+      try {
+        await preAction?.(...actionArgs);
+      } catch (e) {
+        if (e instanceof AbortAction) {
+          return e.callback();
+        } else {
+          throw e;
+        }
+      }
+      const [cmd, argsAndOpts] = getCommandArgsAndOpts(...actionArgs);
+      await promptForMissing(cmd, spec, argsAndOpts);
+      // TODO: normalize network name and include txParams (ConfigManager)
+      Telemetry.report(cmd.name(), argsAndOpts, !!argsAndOpts.interactive);
+      await action(...generateActionArgs(cmd, argsAndOpts));
```

Comment:
Here `argsAndOpts` is an object with key-value pairs of all positional arguments and options, and `generateActionArgs` converts this back into format that commander passes arguments to an action. However, I would consider skipping this conversion and writing our actions to receive a single `argsAndOpts` argument.

Reply:
With some work we can make it check the conversion, but it would approach black magic. Definitely interesting though, TypeScript is super powerful!

Leaving that aside though, I'm not sure that using `argsAndOpts` has stronger type guarantees (unless we also put in that extra work). I just think it would be nicer.